time-manager/System/TimeManager.hs

```diff
@@ -266,3 +268,227 @@ ignoreSync :: E.SomeException -> IO ()
+withHandleKillThread mgr onTimeout action =
+    E.handle ignore $ E.bracket (registerKillThread mgr onTimeout) cancel action
+  where
+    ignore TimeoutThread = return ()
+
+----------------------------------------------------------------
+
+-- | Registering a timeout action.
+register :: Manager -> TimeoutAction -> IO Handle
+register NoManager _ = error "register: NoManager"
+register (Manager timeout) onTimeout = do
+    mgr <- getTimerManager
+    key <- EV.registerTimeout mgr timeout onTimeout
+    keyref <- I.newIORef key
+    let h =
+            Handle
+                { handleTimeout = timeout
+                , handleAction = onTimeout
+                , handleKeyRef = keyref
+                }
+    return h
+
+-- | Unregistering the timeout.
+cancel :: Handle -> IO ()
+cancel Handle{..} = do
+    mgr <- getTimerManager
+    key <- I.readIORef handleKeyRef
+    EV.unregisterTimeout mgr key
+
+-- | Extending the timeout.
+tickle :: Handle -> IO ()
+tickle Handle{..} = do
+    mgr <- getTimerManager
+    key <- I.readIORef handleKeyRef
+#if defined(mingw32_HOST_OS)
+    EV.updateTimeout mgr key $ fromIntegral (handleTimeout `div` 1000000)
+#else
+    EV.updateTimeout mgr key handleTimeout
+#endif
+
+-- | This is identical to 'cancel'.
+--   To resume timeout with the same 'Handle', 'resume' MUST be called.
+--   Don't call 'tickle' for resumption.
+pause :: Handle -> IO ()
+pause = cancel
+
+-- | Resuming the timeout.
+resume :: Handle -> IO ()
+resume Handle{..} = do
+    mgr <- getTimerManager
+    key <- EV.registerTimeout mgr handleTimeout handleAction
+    I.writeIORef handleKeyRef key
+
+----------------------------------------------------------------
+
+-- | The asynchronous exception thrown if a thread is registered via
+-- 'registerKillThread'.
+data TimeoutThread = TimeoutThread
+    deriving (Typeable)
+
+instance E.Exception TimeoutThread where
+    toException = E.asyncExceptionToException
+    fromException = E.asyncExceptionFromException
+
+instance Show TimeoutThread where
+    show TimeoutThread = "Thread killed by timeout manager"
+
+-- | Registering a timeout action of killing this thread.
+--   'TimeoutThread' is thrown to the thread which called this
+--   function on timeout. Catch 'TimeoutThread' if you don't
+--   want to leak the asynchronous exception to GHC RTS.
+registerKillThread :: Manager -> TimeoutAction -> IO Handle
+registerKillThread m onTimeout = do
+    tid <- myThreadId
+    wtid <- mkWeakThreadId tid
+    -- First run the timeout action in case the child thread is masked.
+    register m $
+        onTimeout `E.finally` do
+            mtid <- deRefWeak wtid
+            case mtid of
+                Nothing -> return ()
+                -- FIXME: forkIO to prevent blocking TimerManger
+                Just tid' -> E.throwTo tid' TimeoutThread
+
+----------------------------------------------------------------
+
+-- | Call the inner function with a timeout manager.
+withManager
```

Comment:
The `withManager` and `withManager'` functions are also redundant now, no?
If the `stopManager` and `killManager` functions don't do anything, then why bother wrapping in a bracket?

These can also be removed, potentially